### PR TITLE
`azurerm_data_protection_backup_vault` : support `soft_delete` and `retention_duration_in_days` properties

### PR DESCRIPTION
**Pull request by sinbai**
Swagger : https://github.com/Azure/azure-rest-api-specs/blob/79e4e0c9f8e1c134bcda069fc7994fd99599ccdc/specification/dataprotection/resource-manager/Microsoft.DataProtection/stable/2023-05-01/dataprotection.json#L6970

Test results:
image

### CODE REVIEW

File: website/docs/r/data_protection_backup_vault.html.markdown [hunk 1]
```diff
@@ -47,6 +47,10 @@ The following arguments are supported:
 
 * `identity` - (Optional) An `identity` block as defined below.
 
+* `soft_delete_setting` - (Optional) The state of soft delete for this Backup Vault. Possible values are `AlwaysOn`, `Off` and `On`. Defaults to `On`.
+
+-> **Note:** Once the `soft_delete_setting` is set to `AlwaysOn`, the setting cannot be changed.
```
Comment (mbfrahry) [inline website/docs/r/data_protection_backup_vault.html.markdown:52]: We should add a customize diff that forces a new resource to be created when changing this value from true to false

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Thanks for this PR @sinbai! I am seeing test failures though. Do you mind taking a look at my comment and the tests? 

```
      Error: creating DataProtection BackupVault ("Backup Vault (Subscription: \"1a6092a6-137e-4025-9a7c-ef77f76f2c02\"\nResource Group Name: \"acctest-dataprotection-240205154228526486\"\nBackup Vault Name: \"acctest-bv-240205154228526486\")"): performing CreateOrUpdate: unexpected status 400 with error: BMSUserErrorInvalidInput:  The User input provided for the call is invalid

          with azurerm_data_protection_backup_vault.test,
          on terraform_plugin_test.tf line 28, in resource "azurerm_data_protection_backup_vault" "test":
          28: resource "azurerm_data_protection_backup_vault" "test" {

--- FAIL: TestAccDataProtectionBackupVault_basic (33.15s)
```

File: website/docs/r/data_protection_backup_vault.html.markdown (hunk 1 above)
Comment (sinbai) [reply website/docs/r/data_protection_backup_vault.html.markdown:56]: Fixed.

**Comment by sinbai** [comment]
> Thanks for this PR @sinbai! I am seeing test failures though. Do you mind taking a look at my comment and the tests?
> 
> ```
>       Error: creating DataProtection BackupVault ("Backup Vault (Subscription: \"1a6092a6-137e-4025-9a7c-ef77f76f2c02\"\nResource Group Name: \"acctest-dataprotection-240205154228526486\"\nBackup Vault Name: \"acctest-bv-240205154228526486\")"): performing CreateOrUpdate: unexpected status 400 with error: BMSUserErrorInvalidInput:  The User input provided for the call is invalid
> 
>           with azurerm_data_protection_backup_vault.test,
>           on terraform_plugin_test.tf line 28, in resource "azurerm_data_protection_backup_vault" "test":
>           28: resource "azurerm_data_protection_backup_vault" "test" {
> 
> --- FAIL: TestAccDataProtectionBackupVault_basic (33.15s)
> ```

Hi @mbfrahry thanks for your feedback. I have fixed the above test failure and the comment. Could you please take another look?

File: internal/services/dataprotection/data_protection_backup_vault_resource.go [hunk 2]
```diff
@@ -72,8 +73,35 @@ func resourceDataProtectionBackupVault() *pluginsdk.Resource {
...
+
+		CustomizeDiff: pluginsdk.CustomDiffWithAll(
+			pluginsdk.ForceNewIfChange("soft_delete_setting", func(ctx context.Context, old, new, meta interface{}) bool {
+				return old.(string) == string(backupvaults.SoftDeleteStateAlwaysOn) && new.(string) != string(backupvaults.SoftDeleteStateAlwaysOn)
+			}),
+			pluginsdk.CustomizeDiffShim(func(ctx context.Context, d *pluginsdk.ResourceDiff, v interface{}) error {
+				state := d.Get("soft_delete_setting").(string)
+				_, ok := d.GetOk("retention_duration_in_days")
+				if state == string(backupvaults.SoftDeleteStateOn) && !ok {
+					return fmt.Errorf("`retention_duration_in_days` is required when the `soft_delete_setting` is `On`")
```
Comment (mbfrahry) [inline internal/services/dataprotection/data_protection_backup_vault_resource.go:100]: The way this is implemented will break users as the default is on but `retention_duration_in_days` hasn't been added to the provider yet. So when people update to this version of the provider, Terraform will error with this message. We should probably default `retention_duration_in_days` or remove this check.

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @sinbai, this was close but the way it is implemented will break break users when they update to this version of the provider. We shouldn't change the basic test as that is the bare minimum to get the resource up and running and changing it hid this breaking change.

File: internal/services/dataprotection/data_protection_backup_vault_resource.go (hunk 2 above)
Comment (sinbai) [reply internal/services/dataprotection/data_protection_backup_vault_resource.go]: Hi @mbfrahry thanks a lot. I have updated the code. Could you please take another look?

File: website/docs/r/data_protection_backup_vault.html.markdown [hunk 3]
```diff
@@ -47,6 +47,14 @@ The following arguments are supported:
 
 * `identity` - (Optional) An `identity` block as defined below.
 
+* `retention_duration_in_days` - (Optional) The soft delete retention duration for this Backup Vault. Possible values are between `14` and `180`. Defaults to `14`.
+
+-> **Note:** The `retention_duration_in_days` is the number of days for which deleted data is retained before being permanently deleted. Retention period till 14 days are free of cost, however, retention beyond 14 days may incur additional charges. The `retention_duration_in_days` is required when the `soft_delete_setting` is set to `On`.
+
+* `soft_delete_setting` - (Optional) The state of soft delete for this Backup Vault. Possible values are `AlwaysOn`, `Off` and `On`. Defaults to `On`.
```
Comment (katbyte) [inline website/docs/r/data_protection_backup_vault.html.markdown:54]: i think setting is redundant? as it is implied?
```suggestion
* `soft_delete` - (Optional) The state of soft delete for this Backup Vault. Possible values are `AlwaysOn`, `Off` and `On`. Defaults to `On`.
```

Comment (sinbai) [reply website/docs/r/data_protection_backup_vault.html.markdown]: @katbyte thanks for your feedback. The code has been updated. Could you please take another look?

**Review by katbyte — APPROVED** [review]
LGTM 🗑️

**Pull request merged**